### ISSUE TITLE
Unable to @Override delete methods from PanacheRepositoryBase in an interface

### ISSUE THREAD

**Issue by sacnoth** [quarkusio/quarkus#29296]
### Describe the bug

When I create a repository extending `PanacheRepositoryBase` I want to override the `#deleteAll()` method in my application.  
This works, when I have a repository directly implementing `PanacheRepositoryBase`, but not when I place another interface inbetween the repository for my entity and `PanacheRepositoryBase`.

### Expected behavior

```java
@Table(name = "T_MY_ENTITY")
@Entity
public class MyEntity {
    @Id
    int entityId = 1;

    int entityValue = 2;
}
```

```java
public interface DontDeleteRepo<E, I> extends PanacheRepositoryBase<E, I> {
    String PROHIBIT_DELETE_MESSAGE = "You shall not delete!";

    @Override
    default long deleteAll() {
        throw new UnsupportedOperationException(PROHIBIT_DELETE_MESSAGE);
    }
}
```

```java
@ApplicationScoped
public class MyEntityRepoWithInterface implements DontDeleteRepo<MyEntity, Long> {

}
```

I would expect the overridden `#deleteAll()` method to throw the `UnsupportedOperationException(PROHIBIT_DELETE_MESSAGE)`.

```java
@QuarkusTest
@TestTransaction
class MyEntityRepoWithInterfaceTest {
    @Inject
    MyEntityRepoWithInterface repo;

    @Test
    void deleteAll() {
        assertThatThrownBy(() -> repo.deleteAll())
                .isInstanceOf(UnsupportedOperationException.class)
                .hasMessage(PROHIBIT_DELETE_MESSAGE);
    }
}
```

### Actual behavior

The `UnsupportedOperationException(PROHIBIT_DELETE_MESSAGE)` is not thrown and the delete action is executed.

### How to Reproduce?

I prepared tests that show the problem at [https://github.com/sacnoth/PanacheOverride/](https://github.com/sacnoth/PanacheOverride/).

You an also see there that the issue only exists when an interface like the `DontDeleteRepo` is between the implementing repository `MyEntityRepoWithInterface` and `PanacheRepositoryBase`.

If the `#deleteAll()` method is directly overridden in the implementing repository, as is in the `MyEntityRepoOverride` repository, the `MyEntityRepoOverrideTest` shows that it works as I would expect.

### Output of `uname -a` or `ver`

Windows 10

### Output of `java -version`

OpenJDK Runtime Environment Temurin-17+35 (build 17+35)

### GraalVM version (if different from Java)

none

### Quarkus version or git rev

2.14.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.4

### Additional information

I don't know if other methods besides the `#deleteAll()` are also affected, as I didn't test those.

**Comment by Naganamin** [comment]
Come here after facing the same issue. Is it possible to override methods in `PanacheRepositoryBase` ?

**Comment by FroMage** [comment]
This is likely a bug, it should be possible to override methods.

**Comment by sisimomo** [comment]
Same issue here. Facing the same probleme.

**Comment by sisimomo** [comment]
## 1. The Problem

When you create a base class that **implements** `PanacheRepositoryBase` and then extend it in a final class:
```java
public abstract class BaseRepo<E, ID> implements PanacheRepositoryBase<E, ID> {
    @Override
    public E findById(ID id) {
        // Your custom logic
    }
}

@ApplicationScoped
public class ConcreteRepo extends BaseRepo<MyEntity, UUID> {
    // no override
}
```
you might see that your **base override** (`findById()`) is *ignored* at runtime. Instead, it calls Panache’s default logic.

---

## 2. Why It Happens

- **Quarkus Panache** does **build-time bytecode enhancement** for classes that implement `PanacheRepositoryBase`.
- During this enhancement, Quarkus **generates** the “real” Panache methods (like `find`, `persist`, `delete`) onto the **final** class.
- If your final class has not *explicitly* overridden a Panache method, Quarkus’ generated method in the final class ends up overshadowing any inherited method from your base class that also implements Panache.

Essentially, Quarkus says: 
> “The final class implements Panache but doesn’t override `findById`, so let’s generate our own in `ConcreteRepo`.”

Hence, your base-class override gets replaced at runtime.

---

## 3. The Fix: **Composition** (Hide Panache in Your Base Class)

Instead of making the base class implement `PanacheRepositoryBase`, have it hold a **reference** to the Panache repo and delegate calls:

```java
// (A) Base class: no longer implements Panache
public abstract class ProjectPanacheRepositoryBase<Entity, Id> {

    private PanacheRepositoryBase<Entity, Id> panache;

    void setPanache(PanacheRepositoryBase<Entity, Id> repo) {
        this.panache = repo;
    }

    public Entity findById(ID id) {
        // Your custom logic
    }
}
```

```java
// (B) Final class: implements Panache, injects itself into base
@ApplicationScoped
public class ConcreteRepo 
    extends ProjectPanacheRepositoryBase<MyEntity, UUID>
    implements PanacheRepositoryBase<MyEntity, UUID> {

    @Inject
    void initSelf(ConcreteRepo self) {
        setPanache(self);
    }

    // Now your base class can safely call panache.findById(...)
    // without being overshadowed
}
```

### **Why It Works**

- Your **base class** is now invisible to Quarkus’ Panache enhancer — it does **not** implement `PanacheRepositoryBase`.
- All Panache enhancement happens in the final class (`ConcreteRepo`), which now *does* implement the interface.
- The base class can still perform your “scoped” or “custom” logic by **delegating** to the real Panache methods on the final class reference.

---

## 4. Bottom Line

1. **Don’t** implement Panache in your abstract base class if you want to override them.
2. **Do** implement Panache in your final repo class and delegate to it from your base.
3. This ensures Quarkus’s generated methods don’t overwrite your custom logic.

By hiding Panache behind composition, you keep your “scoping” or other custom logic **fully in control** and unbroken by Quarkus’ build-time magic.

**Comment by Felk** [comment]
Thanks for the detailed workaround (it looks like 100% ChatGPT lol). Is there also a version where the custom repository may still implement `PanacheRepositoryBase` one way or another? It'd be nice to be able to just use the `PanacheRepositoryBase` interface on the consumer side.

**Comment by sisimomo** [comment]
I’ve tried other approaches, but the Panache bytecode enhancement seems really aggressive. I couldn’t find another workaround. That said, I’m far from an expert with Panache—I only started using it about five months ago. Still, I’m pretty happy with the workaround.

P.S. Yep, the text is ChatGPT—team effort ;) I’m the brain, and he does the writing.

**Comment by celcius112** [comment]
@FroMage any news on this issue ? This is a very limiting and insidious bug.

**Comment by FroMage** [comment]
No news, no, but it should be fairly easy to fix. I can point you to where to fix it if you feel up for providing a PR?

**Comment by celcius112** [comment]
@FroMage sure ! that could be interesting

**Comment by FroMage** [comment]
Great!

So, we currently avoid user-defined overrides by collecting their names and signatures in https://github.com/quarkusio/quarkus/blob/main/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/PanacheRepositoryClassOperationGenerationVisitor.java#L117 and avoiding generating overrides at https://github.com/quarkusio/quarkus/blob/main/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/PanacheRepositoryClassOperationGenerationVisitor.java#L126

Note that I've verified that this currently ignores `default` interface methods as well as any override defined in any superclass between the repository and `PanacheRepositoryBase`, so the problem exists not just for interfaces, but also regular superclasses.

Ideally, in `visit(int version, int access, String name, String signature, String superName, String[] interfaces)` we'd use `indexView` to visit all superclasses and superinterfaces going up to `panacheRepositoryBaseClassInfo` and collect all the overridden methods in `userMethods`, as long as the methods are not abstract.

This should seed the set of methods we should not override.

I suggest starting with two failing tests in `integration-tests/hibernate-orm-panache` with a default method in a superinterface, and a concrete method in a superclass, then work towards fixing the tests.

Then add abstract methods in the superclass and superinterface to verify that those are not taken into account.

Once that is done, we can already call this progress and merge and postpone the tricky cases for later issues: generics, kotlin and hibernate-reactive. Or we can also do them if you're motivated :)
